NN_closure_global_nof_BBLintegral.jl

```diff
@@ -0,0 +1,224 @@
+end
+
+@inline function find_based_boundary_layer_index!(i, j, field, grid, h)
+    kmax = 3
+
+    @inbounds for k in 6:grid.Nz-1
+        kmax = ifelse(field[i, j, k] > 0, k-2, kmax)
+    end
+
+    @inbounds h[i, j, 1] = kmax
+end
+
+@kernel function _find_base_boundary_layer!(BBL_constraint, grid, h)
+    i, j = @index(Global, NTuple)
+    find_based_boundary_layer_index!(i, j, BBL_constraint, grid, h)
+end
+
+@kernel function _adjust_nn_fluxes!(diffusivities, grid, closure::NNFluxClosure, tracers, velocities, buoyancy, top_tracer_bcs, clock)
+    i, j, k = @index(Global, NTuple)
+    scaling = closure.scaling
+    @inbounds BBL_index = diffusivities.BBL_index[i, j, 1]
+
+    convecting = top_buoyancy_flux(i, j, grid, buoyancy, top_tracer_bcs, clock, merge(velocities, tracers)) > 0
+    above_base_boundary_layer = k > BBL_index
+    below_top = k <= grid.Nz - 1
+    above_bottom = k >= 3
+
+    NN_active = convecting & above_base_boundary_layer & below_top & above_bottom
+
+    @inbounds diffusivities.wT[i, j, k] = ifelse(NN_active, scaling.wT.σ * diffusivities.wT[i, j, k], 0)
+    @inbounds diffusivities.wS[i, j, k] = ifelse(NN_active, scaling.wS.σ * diffusivities.wS[i, j, k], 0)
+end
+
+# Write here your constructor
+# NNFluxClosure() = ... insert NN here ... (make sure it is on GPU if you need it on GPU!)
+
+const NNC = NNFluxClosure
+                                                         
+#####
+##### Abstract Smagorinsky functionality
+#####
+
+# Horizontal fluxes are zero!
+@inline viscous_flux_wz( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_wx( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_wy( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_ux( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_vx( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_uy( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_vy( i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline diffusive_flux_x(i, j, k, grid, clo::NNC, K, ::Val{tracer_index}, c, clock, fields, buoyancy) where tracer_index = zero(grid)
+@inline diffusive_flux_y(i, j, k, grid, clo::NNC, K, ::Val{tracer_index}, c, clock, fields, buoyancy) where tracer_index = zero(grid)
+
+# Viscous fluxes are zero (for now)
+@inline viscous_flux_uz(i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+@inline viscous_flux_vz(i, j, k, grid, clo::NNC, K, clk, fields, b) = zero(grid)
+
+# The only function extended by NNFluxClosure
+@inline diffusive_flux_z(i, j, k, grid, clo::NNC, K, ::Val{1}, c, clock, fields, buoyancy) = @inbounds K.wT[i, j, k]
+@inline diffusive_flux_z(i, j, k, grid, clo::NNC, K, ::Val{2}, c, clock, fields, buoyancy) = @inbounds K.wS[i, j, k]
```

Comment:
seems like it would be better to have the tracer names here since `tracers=(:S, :T)` would make this fail right?

It's possible to refactor diffusive fluxes for that, eg we already use names for the bgc:

https://github.com/CliMA/Oceananigans.jl/blob/d367f112e0ae5c279fe3a63f1726469540d4eb5b/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl#L136